File: benchmark/c/opl/opl_loadbank_ibk/opl_loadbank_ibk.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define OPL_EMU_REGISTERS_OPERATORS ( OPL_EMU_REGISTERS_CHANNELS * 2 )
#define OPL_EMU_REGISTERS_WAVEFORMS 8
#define OPL_EMU_REGISTERS_CHANNELS 18
#define OPL_EMU_REGISTERS_REGISTERS 0x200
#define OPL_EMU_REGISTERS_WAVEFORM_LENGTH 0x400

enum opl_emu_envelope_state
{
	OPL_EMU_EG_ATTACK = 1,
	OPL_EMU_EG_DECAY = 2,
	OPL_EMU_EG_SUSTAIN = 3,
	OPL_EMU_EG_RELEASE = 4,
	OPL_EMU_EG_STATES = 6
};
enum op2_flags_t {
  OP2_FIXEDPITCH = 1,
  OP2_UNUSED = 2,
  OP2_DOUBLEVOICE = 4,
};
typedef struct opl_t opl_t;


struct opl_emu_registers
{
	uint16_t m_lfo_am_counter;
	uint16_t m_lfo_pm_counter;
	uint32_t m_noise_lfsr;
	uint8_t m_lfo_am;
	uint8_t m_regdata[OPL_EMU_REGISTERS_REGISTERS];
	uint16_t m_waveform[OPL_EMU_REGISTERS_WAVEFORMS][OPL_EMU_REGISTERS_WAVEFORM_LENGTH];
};

struct opl_emu_opdata_cache
{
	uint32_t phase_step;
	uint32_t total_level;
	uint32_t block_freq;
	int32_t detune;
	uint32_t multiple;
	uint32_t eg_sustain;
	uint8_t eg_rate[OPL_EMU_EG_STATES];
	uint8_t eg_shift;
};
struct opl_emu_fm_operator
{
	uint32_t m_choffs;
	uint32_t m_opoffs;
	uint32_t m_phase;
	uint16_t m_env_attenuation;
	enum opl_emu_envelope_state m_env_state;
	uint8_t m_key_state;
	uint8_t m_keyon_live;
	struct opl_emu_opdata_cache m_cache;
	struct opl_emu_registers* m_regs;
};
struct opl_emu_fm_channel
{
	uint32_t m_choffs;
	int16_t m_feedback[2];
	int16_t m_feedback_in;
};
typedef struct opl_timbre_t {
  unsigned long modulator_E862, carrier_E862;
  unsigned char modulator_40, carrier_40;
  unsigned char feedconn;
  signed char finetune;
  unsigned char notenum;
  signed short noteoffset;
} opl_timbre_t;
struct opl_emu_t
{
	uint32_t m_env_counter;
	uint8_t m_status;
	uint8_t m_timer_running[2];
	uint32_t m_active_channels;
	uint32_t m_modified_channels;
	uint32_t m_prepare_count;
	struct opl_emu_registers m_regs;
	struct opl_emu_fm_channel m_channel[OPL_EMU_REGISTERS_CHANNELS];
	struct opl_emu_fm_operator m_operator[OPL_EMU_REGISTERS_OPERATORS];
};
struct voicealloc_t {
  unsigned short priority;
  signed short timbreid;
  signed char channel;
  signed char note;
  unsigned char voiceindex;
};
struct opl_t {
  signed char notes2voices[16][128][2];
  unsigned short channelpitch[16];
  unsigned short channelvol[16];
  struct voicealloc_t voices2notes[18];
  unsigned char channelprog[16];
  int opl3;
  struct opl_emu_t opl_emu;
  struct opl_timbre_t opl_gmtimbres[ 256 ];
  struct opl_timbre_t opl_gmtimbres_voice2[ 256 ];
  int is_op2;
  enum op2_flags_t op2_flags[ 256 ];
};
static int opl_loadbank_internal(opl_t* opl, char const* file, int offset) ;
int opl_loadbank_ibk(opl_t* opl, char const* file) ;
static int opl_loadbank_internal(opl_t* opl, char const* file, int offset) {
  opl->is_op2 = 0;
  unsigned char buff[16];
  int i;
  FILE* f = fopen( file, "rb" );
  if( !f ) return -1;
  fseek( f, 0, SEEK_END );
  if (ftell(f) != 3204) {
    fclose(f);
    return(-2);
  }
  fseek( f, 0, SEEK_SET);
  if ((fread(buff, 1, 4,f) != 4) || (buff[0] != 'I') || (buff[1] != 'B') || (buff[2] != 'K') || (buff[3] != 0x1A)) {
    fclose(f);
    return(-3);
  }
  for (i = offset; i < 128 + offset; i++) {
    if (fread(buff, 1, 16, f) != 16) {
      fclose(f);
      return(-4);
    }
    opl->opl_gmtimbres[i].modulator_E862 = buff[8];
    opl->opl_gmtimbres[i].modulator_E862 <<= 8;
    opl->opl_gmtimbres[i].modulator_E862 |= buff[6];
    opl->opl_gmtimbres[i].modulator_E862 <<= 8;
    opl->opl_gmtimbres[i].modulator_E862 |= buff[4];
    opl->opl_gmtimbres[i].modulator_E862 <<= 8;
    opl->opl_gmtimbres[i].modulator_E862 |= buff[0];
    opl->opl_gmtimbres[i].carrier_E862 = buff[9];
    opl->opl_gmtimbres[i].carrier_E862 <<= 8;
    opl->opl_gmtimbres[i].carrier_E862 |= buff[7];
    opl->opl_gmtimbres[i].carrier_E862 <<= 8;
    opl->opl_gmtimbres[i].carrier_E862 |= buff[5];
    opl->opl_gmtimbres[i].carrier_E862 <<= 8;
    opl->opl_gmtimbres[i].carrier_E862 |= buff[1];
    opl->opl_gmtimbres[i].modulator_40 = buff[2];
    opl->opl_gmtimbres[i].carrier_40 = buff[3];
    opl->opl_gmtimbres[i].feedconn = buff[10];
    opl->opl_gmtimbres[i].finetune = buff[12];
    opl->opl_gmtimbres[i].notenum = 60;
    opl->opl_gmtimbres[i].noteoffset = 0;
  }
  fclose(f);
  return(0);
}
int opl_loadbank_ibk(opl_t* opl, char const* file) {
  char *instruments = NULL, *percussion = NULL;
  int i, res;
  instruments = strdup(file);
  if (instruments == NULL) return(-64);
  for (i = 0; instruments[i] != 0; i++) {
    if (instruments[i] == ',') {
      instruments[i] = 0;
      percussion = instruments + i + 1;
      break;
    }
  }
  res = opl_loadbank_internal(opl, instruments, 0);
  if ((res == 0) && (percussion != NULL)) {
    res = opl_loadbank_internal(opl, percussion, 128);
  }
  free(instruments);
  return(res);
}

File: flourine_reslut/fail/c/opl/opl_loadbank_ibk/claude2/CAPR/opl_loadbank_ibk.rs


use std::fs::File;
use std::io::{Read, Seek, SeekFrom};
use std::boxed::Box;
use std::vec::Vec;

const OPL_EMU_REGISTERS_OPERATORS: usize = OPL_EMU_REGISTERS_CHANNELS * 2;

const OPL_EMU_REGISTERS_WAVEFORMS: usize = 8;

const OPL_EMU_REGISTERS_CHANNELS: usize = 18;

const OPL_EMU_REGISTERS_REGISTERS: usize = 0x200;

const OPL_EMU_REGISTERS_WAVEFORM_LENGTH: usize = 0x400;

enum OplEmuEnvelopeState {
    OplEmuEgAttack = 1,
    OplEmuEgDecay = 2,
    OplEmuEgSustain = 3,
    OplEmuEgRelease = 4,
    OplEmuEgStates = 6,  
}

enum Op2Flags {
    Op2Fixedpitch = 1,
    Op2Unused = 2,
    Op2Doublevoice = 4
}

struct OplEmuRegisters {
    m_lfo_am_counter: u16,
    m_lfo_pm_counter: u16,
    m_noise_lfsr: u32,
    m_lfo_am: u8,
    m_regdata: [u8; OPL_EMU_REGISTERS_REGISTERS],
    m_waveform: [[u16; OPL_EMU_REGISTERS_WAVEFORM_LENGTH]; OPL_EMU_REGISTERS_WAVEFORMS],
}

struct OplEmuOpdataCache {
    phase_step: u32,
    total_level: u32,
    block_freq: u32,
    detune: i32,
    multiple: u32,
    eg_sustain: u32,
    eg_rate: [u8; OplEmuEnvelopeState::OplEmuEgStates as usize],
    eg_shift: u8
}

struct OplEmuFmOperator {
    m_choffs: u32,
    m_opoffs: u32,
    m_phase: u32,
    m_env_attenuation: u16,
    m_env_state: OplEmuEnvelopeState,
    m_key_state: u8,
    m_keyon_live: u8,
    m_cache: Box<OplEmuOpdataCache>
}

struct OplEmuFmChannel {
    m_choffs: u32,
    m_feedback: [i16; 2],
    m_feedback_in: i16    
}

struct OplTimbre {
    modulator_E862: u32,
    carrier_E862: u32,
    modulator_40: u8,
    carrier_40: u8,
    feedconn: u8,
    finetune: i8,
    notenum: u8,
    noteoffset: i16
}

struct OplEmu;

struct Opl {
    notes2voices: [[[[i8; 2]; 128]; 16]; 2],
    channelpitch: [u16; 16],   
    channelvol: [u16; 16],
    voices2notes: [Box<VoiceAlloc>; 18],
    channelprog: [u8; 16],
    opl3: i32,
    opl_emu: Box<OplEmu>,
    opl_gmtimbres: [Box<OplTimbre>; 256],
    opl_gmtimbres_voice2: [Box<OplTimbre>; 256],  
    is_op2: bool,
    op2_flags: [Op2Flags; 256]
}

struct VoiceAlloc {
    priority: u16,
    timbreid: i16,
    channel: i8,
    note: i8,
    voiceindex: u8  
}

fn opl_loadbank_internal(opl: &mut Opl, file: &str, offset: i32) -> i32 {
    offset
}

fn opl_loadbank_ibk(opl: &mut Opl, file: &str) -> i32 {
    0 // return 0
}




